File: Python/0032_longest_valid_parentheses/0032_longest_valid_parentheses.py
class Solution:
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) < 2:
            return 0
        max_value = 0
        dp = [0] * len(s)
        for i in range(1, len(s)):
            if (s[i] == ')'):
                if (s[i - 1] == '('):
                    dp[i] = (dp[i - 2] if i >= 2 else 0) + 2
                elif (i - dp[i - 1] > 0 and s[i - dp[i - 1] - 1] == '('):
                    dp[i] = dp[i - 1] + (dp[i - dp[i - 1] - 2] if (i - dp[i - 1]) >= 2 else 0) + 2
            max_value = max(max_value, dp[i]);
        return max_value

File: C/0032_longest_valid_parentheses/0032_longest_valid_parentheses.c
int longestValidParentheses(char* s) {
    int max = 0;
    int len = strlen(s);
    if (len < 2) {
        return 0;
    }
    int *dp = malloc(len * sizeof(int));
    memset(dp, 0, sizeof(int) * len);
    for (int i = 1; i < len; i++) {
        if (s[i] == ')') {
            if (s[i - 1] == '(') {
                dp[i] = (i >= 2 ? dp[i - 2] : 0) + 2;
            }
            else if (i - dp[i - 1] > 0 && s[i - dp[i - 1] - 1] == '(') {
                dp[i] = dp[i - 1] + ((i - dp[i - 1]) >= 2 ? dp[i - dp[i - 1] - 2] : 0) + 2;
            }
            max = max > dp[i] ? max : dp[i];
        }
    }
    free(dp);
    return max;
}